[Cocoa] Pasting a Genmoji into Mail does not create a HEIC attachment

pxlcoder opened the pull request:
#### 0ed894383eb476f3a6a2abda84a5f71c06d48c58
<pre>
[Cocoa] Pasting a Genmoji into Mail does not create a HEIC attachment
<a href="https://bugs.webkit.org/show_bug.cgi?id=275982">https://bugs.webkit.org/show_bug.cgi?id=275982</a>
<a href="https://rdar.apple.com/129805233">rdar://129805233</a>

Reviewed by Abrar Rahman Protyasha.

When dragging or copying an image element, attachment elements are preserved in
dropped/pasted fragment by writing a temporary &quot;webkitattachmentid&quot; attribute to
the serialized markup on copy. Then, upon deserialization an attachment element
with the same identifier is created under the image.

However, this logic was not being applied to &lt;source&gt; elements, which contain the
data necessary to render a Genmoji. This resulted in Genmoji being lost, leaving
behind only a fallback PNG, when copying from another app, pasting into Mail and
then sending an email. No issues were observed simply when pasting, as the &lt;source&gt;
element was still created. However, Mail is never notified of a new attachment, so
their serialization upon sending fails to capture the full fidelity Genmoji.

Fix by generalizing the approach used to preserve image attachments, to
`AttachmentAssociatedElement`, which includes &lt;image&gt; and &lt;source&gt;.

* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::appendCustomAttributes):
(WebCore::restoreAttachmentElementsInFragment):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertWKAttachmentsOnPaste)):

Canonical link: <a href="https://commits.webkit.org/280459@main">https://commits.webkit.org/280459@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/477db229377e98deebf16fc820a41e8890c41079

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56700 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36027 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9173 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60314 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7141 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58827 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7332 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45935 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5012 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58730 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33865 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48942 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26791 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30648 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6273 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6144 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52602 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6544 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61997 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/612 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/6643 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53196 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/614 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48997 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53214 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12543 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/527 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31857 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32941 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34029 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32687 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
```diff
@@ -394,6 +394,68 @@ - (NSAttributedString *)contentsAsAttributedString
+    RetainPtr font = [NSFont systemFontOfSize:16];
+#else
+    RetainPtr font = [UIFont systemFontOfSize:16];
+#endif
```
We could save some typing here: `RetainPtr font = [WebCore::CocoaFont systemFontOfSize:16];`

pxlcoder committed:
[Cocoa] Pasting a Genmoji into Mail does not create a HEIC attachment
https://bugs.webkit.org/show_bug.cgi?id=275982
rdar://129805233

Reviewed by Abrar Rahman Protyasha.

When dragging or copying an image element, attachment elements are preserved in
dropped/pasted fragment by writing a temporary "webkitattachmentid" attribute to
the serialized markup on copy. Then, upon deserialization an attachment element
with the same identifier is created under the image.

However, this logic was not being applied to <source> elements, which contain the
data necessary to render a Genmoji. This resulted in Genmoji being lost, leaving
behind only a fallback PNG, when copying from another app, pasting into Mail and
then sending an email. No issues were observed simply when pasting, as the <source>
element was still created. However, Mail is never notified of a new attachment, so
their serialization upon sending fails to capture the full fidelity Genmoji.

Fix by generalizing the approach used to preserve image attachments, to
`AttachmentAssociatedElement`, which includes <image> and <source>.

* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::appendCustomAttributes):
(WebCore::restoreAttachmentElementsInFragment):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertWKAttachmentsOnPaste)):

Canonical link: https://commits.webkit.org/280459@main
